rust/kernel/file_operations.rs

```diff
@@ -79,13 +79,26 @@ pub enum SeekFrom {
     Current(i64),
 }
 
+/// Called by the VFS when an inode should be opened.
+///
+/// Calls `T::open` on the returned value of `A::convert`.
+///
+/// # Safety
+///
+/// The returned value of `A::convert` must be a valid non-null pointer and
+/// `T:open` must return a valid non-null pointer on an `Ok` result.
 unsafe extern "C" fn open_callback<A: FileOpenAdapter, T: FileOpener<A::Arg>>(
     inode: *mut bindings::inode,
     file: *mut bindings::file,
 ) -> c_types::c_int {
     from_kernel_result! {
+        // SAFETY: `A::convert` must return a valid non-null pointer
         let arg = unsafe { A::convert(inode, file) };
+        // SAFETY: `arg` was returned by `A::convert` and must be a valid
+        // non-null pointer
         let ptr = T::open(unsafe { &*arg })?.into_pointer();
+        // SAFETY: `file` was returned by `T::open` and must be a valid
+        // non-null pointer.
         unsafe { (*file).private_data = ptr as *mut c_types::c_void };
```

Comment:
To arrive at a correct comment here, ask yourself how you can prove that it is **always** safe to dereference the raw pointer called `file`. Is that pointer guaranteed to be valid and non-null? If so, why?

(note that I do not know the answer to that question myself, this is intended to show what questions to ask and answer)

Reply:
Updated this one a bit. I whiffed on the first draft.